plugins/interactive/shellserver.go

```diff
@@ -0,0 +1,192 @@
+package interactive
+
+import (
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/gorilla/websocket"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
+)
+
+const (
+	// ShellHandshakeTimeout is the maximum allowed time for websocket handshake
+	ShellHandshakeTimeout = 30 * time.Second
+	// ShellPingInterval is the time between sending pings
+	ShellPingInterval = 15 * time.Second
+	// ShellWriteTimeout is the maximum time between successful writes
+	ShellWriteTimeout = ShellPingInterval * 2
+	// ShellPongTimeout is the maximum time between successful reads
+	ShellPongTimeout = ShellPingInterval * 3
+	// ShellBlockSize is the maximum number of bytes to send in a single block
+	ShellBlockSize = 16 * 1024
+	// ShellMaxMessageSize is the maximum message size we will read
+	ShellMaxMessageSize = ShellBlockSize + 4*1024
+	// ShellMaxPendingBytes is the maximum number of bytes allowed in-flight
+	ShellMaxPendingBytes = 4 * ShellBlockSize
+)
+
+// ShellFactory is a function that can make a shell
+type ShellFactory func(command []string, tty bool) (engines.Shell, error)
+
+// A ShellServer implements http.Handler and upgrades connections to websockets,
+// creates a shell and connects the websocket to the shell.
+type ShellServer struct {
+	m             sync.Mutex
+	c             sync.Cond
+	makeShell     ShellFactory
+	done          chan struct{}
+	refCount      int
+	instanceCount int
+	log           *logrus.Entry
+}
+
+// NewShellServer returns a new ShellServer which creates shells using the
+// makeShell function.
+func NewShellServer(makeShell ShellFactory, log *logrus.Entry) *ShellServer {
+	s := &ShellServer{
+		makeShell: makeShell,
+		done:      make(chan struct{}),
+		log:       log,
+	}
+	s.c.L = &s.m
+	return s
+}
+
+// Wait will wait for all active shells to be done and return
+func (s *ShellServer) Wait() {
+	s.m.Lock()
+	defer s.m.Unlock()
```

Comment:
What are we protecting here? `refCount`? If so, how is its value supposed to reach zero if the mutex is hold in the `for` iteration?